src/foraging_gui/TransferToNWB.py

```diff
@@ -271,42 +289,45 @@ class obj:
-            LaserLocationC = 0
-            LaserPowerC = 0
-            LaserDurationC = 0
-            LaserConditionC = 0
-            LaserConditionProC = 0
-            LaserStartC = 0
-            LaserStartOffsetC = 0
-            LaserEndC = 0
-            LaserEndOffsetC = 0
-            LaserProtocolC = 0
-            LaserFrequencyC = 0
-            LaserRampingDownC = 0
-            LaserPulseDurC = 0
+            LaserWavelengthC = np.nan
+            LaserLocationC = 'None'
+            Laser1Power = np.nan
+            Laser2Power = np.nan
+            LaserOnProbablityC = np.nan
+            LaserDurationC = np.nan
+            LaserConditionC = 'None'
+            LaserConditionProC = np.nan
+            LaserStartC = 'None'
+            LaserStartOffsetC = np.nan
+            LaserEndC = 'None'
+            LaserEndOffsetC = np.nan
+            LaserProtocolC = 'None'
+            LaserFrequencyC = np.nan
+            LaserRampingDownC = np.nan
+            LaserPulseDurC = np.nan
+
         else:
-            # if there is no training paramters history stored (for old Bonsai behavior control)
-            if not hasattr(obj, 'TP_Laser_1'):
-                if getattr(obj, f'TP_Laser_{Sc}')[i] == 'Blue':
-                    LaserWavelengthC = 473
-                elif getattr(obj, f'TP_Laser_{Sc}')[i] == 'Red':
-                    LaserWavelengthC = 647
-                elif getattr(obj, f'TP_Laser_{Sc}')[i] == 'Green':
-                    LaserWavelengthC = 547
-                LaserLocationC = getattr(obj, f'TP_Location_{Sc}')[i]
-                LaserPowerC = getattr(obj, f'TP_LaserPower_{Sc}')[i]
-                LaserDurationC = getattr(obj, f'TP_Duration_{Sc}')[i]
-                LaserConditionC = getattr(obj, f'TP_Condition_{Sc}')[i]
-                LaserConditionProC = getattr(obj, f'TP_ConditionP_{Sc}')[i]
-                LaserStartC = getattr(obj, f'TP_LaserStart_{Sc}')[i]
-                LaserStartOffsetC = getattr(obj, f'TP_OffsetStart_{Sc}')[i]
-                LaserEndC = getattr(obj, f'TP_LaserEnd_{Sc}')[i]
-                LaserEndOffsetC = getattr(obj, f'TP_OffsetEnd_{Sc}')[i]
-                LaserProtocolC = getattr(obj, f'TP_Protocol_{Sc}')[i]
-                LaserFrequencyC = getattr(obj, f'TP_Frequency_{Sc}')[i]
-                LaserRampingDownC = getattr(obj, f'TP_RD_{Sc}')[i]
-                LaserPulseDurC = getattr(obj, f'TP_PulseDur_{Sc}')[i]
+            if getattr(obj, f'TP_Laser_{Sc}')[i] == 'Blue':
+                LaserWavelengthC = float(473)
+            elif getattr(obj, f'TP_Laser_{Sc}')[i] == 'Red':
+                LaserWavelengthC = float(647)
+            elif getattr(obj, f'TP_Laser_{Sc}')[i] == 'Green':
+                LaserWavelengthC = float(547)
+            LaserLocationC = str(getattr(obj, f'TP_Location_{Sc}')[i])
+            Laser1Power=float(eval(_get_field(obj, field_list=[f'TP_Laser1_power_{Sc}',f'TP_LaserPowerLeft_{Sc}'],index=i,default='[np.nan,np.nan]'))[1])
```

Comment:
I have a new question about this. Is this backward compatible? The previous code is
```python
LaserPowerC = getattr(obj, f'TP_LaserPower_{Sc}')[i]
```
but now your `field_list` becomes `field_list=[f'TP_Laser1_power_{Sc}', f'TP_LaserPowerLeft_{Sc}']`. Shouldn't the second option be `f'TP_LaserPower_{Sc}'`?

Reply:
Hi Han, the second field should be `f'TP_LaserPowerLeft_{Sc}'` instead of  `f'TP_LaserPower_{Sc}'`. In the initial code, I did not differentiate Left and Right laser, so it was `f'TP_LaserPower_{Sc}'`. We don't have opto sessions with `f'TP_LaserPower_{Sc}'`.

Reply:
I don't understand. You mean previously it has always been f'TP_LaserPowerLeft_{Sc}' in the json file, and the one you used f'TP_LaserPower_{Sc}' was actually wrong?

Reply:
In the very beginning version of the GUI, the left and right lasers were combined in f'TP_LaserPower_{Sc}' just like your previous GUI. I then separated them to  f'TP_LaserPowerLeft_{Sc}' and f'TP_LaserPowerRight_{Sc}'.

Reply:
My question is: has the current code correctly captured all historical field names of laser power?

From what you're describing:

Phase 1. You start with f'TP_LaserPower_{Sc}' in both json and convertNWB code

Phase 2. Then at some point you split it to left and right in bonsai json, but you didn't change in the convertNWB code (that is why I asked whether the convert code was "wrong" during that period of time)

Phase 3. In the recent update of the GUI, you changed the laser power name from left/right to 1/2. So in this PR, you add the new name with 1/2 in field_list, together with a left/right for backward compatibility, which finally fixes the mismatch in phase 2.

Are I understanding correctly?

If so, have you ever collected photostim data during phase 1? If you did, then the current PR is not compatible with those data.